skrub/datasets/_utils.py

```diff
@@ -54,3 +81,155 @@ def get_data_dir(name=None, data_home=None):
+
+    bunch = Bunch()
+    for file_path in datafiles_dir.iterdir():
+        if file_path.suffix == ".csv":
+            bunch[file_path.stem] = pd.read_csv(file_path)
+        elif file_path.suffix == ".json":
+            bunch[file_path.stem] = json.loads(file_path.read_text(encoding="utf-8"))
+
+    return bunch
+
+
+def _extract_archive(dataset_dir, archive_path, diff_checksum):
+    dataset_name = dataset_dir.name
+
+    if not archive_path.exists() or diff_checksum:
+        _download_archive(dataset_name, archive_path)
+
+    shutil.unpack_archive(archive_path, dataset_dir, format="zip")
+
+
+def _download_archive(
+    dataset_name,
+    archive_path,
+    retry=3,
+    delay=1,
+    timeout=30,
+    chunk_size=4096,
+):
+    metadata = DATASET_INFO[dataset_name]
+
+    for idx in range(1, retry + 1):
+        for target_url in metadata["urls"]:
+            print(
+                f"Downloading {dataset_name!r} from {target_url} (retry {idx}/{retry})"
+            )
+            try:
+                error_flag = False
+                _stream_download(archive_path, target_url, timeout, chunk_size)
+            except Exception as e:
+                error_flag = True
+                warnings.warn(repr(e), category=FutureWarning)
+
+            if not error_flag:
+                if _sha256(archive_path) != metadata["sha256"]:
+                    raise OSError(
+                        f"File {archive_path.stem!r} checksum verification has failed, "
+                        "which means the remote file has been updated.\n"
+                        "Please update your skrub version."
+                    )
+                return
+        time.sleep(delay)
+        delay *= 5
+
+    else:
+        raise OSError(
+            f"Can't download the file {dataset_name!r} from urls {metadata['urls']}."
+        )
+
+
+def _stream_download(
+    archive_path,
+    target_url,
+    timeout,
+    chunk_size,
+):
+    # We create a temporary file dedicated to this particular download to avoid
+    # conflicts with parallel downloads. If the download is successful, the
+    # temporary file is atomically renamed to the final file path (with
+    # `shutil.move`). We therefore pass `delete=False` to `NamedTemporaryFile`.
+    # Otherwise, garbage collecting temp_file would raise an error when
+    # attempting to delete a file that was already renamed. If the download
+    # fails, the temporary file is removed manually in the except block.
+    temp_file = NamedTemporaryFile(
+        mode="wb",
+        prefix=archive_path.stem + ".part_",
+        dir=archive_path.parent,
+        delete=False,
+    )
+
+    try:
+        temp_file_path = Path(temp_file.name)
+        response = requests.get(target_url, timeout=timeout, stream=True)
+        for chunk in response.iter_content(chunk_size):
+            temp_file.write(chunk)
+
+    except (Exception, KeyboardInterrupt):
+        Path(temp_file.name).unlink()
```

Comment:
here I don't think we have closed the file so on windows trying to remove it will raise an exception. maybe the easiest thing is to close the file right after creating it and then open it again in a `with` statement